### PR TITLE
Resolve correct link path for tsconfig.extends

### LINKED ISSUE #131643: VSCode unable to resolve `extends: "my-npm-tsconfig"`, requires an extra `/tsconfig.json`
Issue Type: <b>Bug</b>

TypeScript can resolve this configuration correctly but VSCode can't

```json
{
	"extends": "@sindresorhus/tsconfig"
}
```

Clicking it brings up this VSCode error

> Unable to read file './node_modules/@sindresorhus/tsconfig.json' (Error: Unable to resolve non-existing file './node_modules/@sindresorhus/tsconfig.json')

The following is clickable in VSCode, but it shouldn't be necessary:

```json
{
	"extends": "@sindresorhus/tsconfig/tsconfig.json"
}
```

npm package: https://github.com/sindresorhus/tsconfig/blob/640e4bfe1172e27d59bf4d73a2c5df517cc01e1d/package.json#L12

Side note: Both configurations are clickable in WebStorm

VS Code version: Code 1.59.1 (3866c3553be8b268c8a7f8c0482c0c0177aa8bfa, 2021-08-19T11:53:52.479Z)
OS version: Darwin x64 20.6.0
Restricted Mode: No

<details><summary>Extensions (18)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-css-formatter|aes|1.0.1
sidebar-markdown-notes|ass|1.0.4
vscode-fish|bma|1.0.24
prettier-vscode-standard|bys|0.22.1
vscode-mac-color-picker|dae|1.0.2
vscode-eslint|dba|2.1.23
prettier-vscode|esb|8.1.0
codespaces|Git|1.0.3
remotehub|Git|0.13.4
vscode-github|Kni|0.30.6
dotenv|mik|1.0.1
sublime-keybindings|ms-|4.0.9
ghosttext|ngy|0.0.2
docthis|oou|0.8.2
vscode-fileutils|sle|3.4.5
svelte-vscode|sve|105.3.5
tabnine-vscode|Tab|3.4.22
shellcheck|tim|0.15.2

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
pythonvspyt678:30270856
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30300192
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vsdfh931cf:30280410
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
pythonvsuse255:30340121
vscod805:30301674
pythonvspyt200:30340761
vscextlang:30333561
binariesv615:30325510
pythonvssor306:30344512
bridge0708:30335490
vstre464cf:30350173
bridge0723:30353136
javagetstartedc:30351792
vsdyn478:30356813

```

</details>

### LINKED ISSUE #141112: Incorrect `extends` resolving for tsconfig
Currectly, tsconfig `extends` links are hardcoded and don't resolve to correct path comparing to TS behavior. TypeScript algorithm reference: https://github.com/microsoft/TypeScript/blob/febfd442cdba343771f478cf433b0892f213ad2f/src/compiler/commandLineParser.ts#L3005.

I checked it with `ts --showConfig` command.

### PR DESCRIPTION
**Pull request by zardoy**
This PR fixes #141112

@mjbvz Sorry for the latency, but its here. I tried, but failed to reuse [existing resolve algorithm](https://github.com/microsoft/TypeScript/blob/febfd442cdba343771f478cf433b0892f213ad2f/src/compiler/commandLineParser.ts#L3005). However I hope it would be possible in future with standalone lib like `tsconfig-paths`.

I think this is almost 1:1 implementation to what we have in TypeScript repo today. Now, if the file doesn't exist, it won't create a link to a non-existing file.
I didin't handle a case where tsconfig might be referenced from the root since its a rare case and I won't believe someone actually does that (however we can check whether path starts with `/` at least).

I made everything up to the provider method async only because of async `exists` (`fs.stat`).

@fregante I remember you compared it to WebStorm. This implementation is much better!

However, I noticed a bug that if you write path to the existing file and then swap it with a non-existing, the provider would be called and it'd return `undefined`, but the old link wouldn't be dropped.

### CODE REVIEW

**Comment by zardoy** [comment]
I just remembered that I forgot to describe purpose of `resolveNodeModulesPath`. Here is example with library from linked issue:
```txt
- 📁 node_modules
    - 📁 @sindresorhus
        - 📁 tsconfig
            - tsconfig <-- would resolve to here!
- packages/some-package
    - 📁 node_modules
        - 📁 @sindresorhus
            - 📁 another-lib
    - 📁 lib
        - tsconfig.json (extends: @sindresorhus/tsconfig)
```

File: extensions/typescript-language-features/src/utils/fs.ts [hunk 1]
```diff
@@ -14,3 +15,31 @@ export const exists = async (resource: vscode.Uri): Promise<boolean> => {
 		return false;
 	}
 };
+
+/** A lighter Node's node_modules resolution algorithm */
+export const resolveNodeModulesPath = async (currentUriDir: vscode.Uri, pathCandidates: string[]): Promise<vscode.Uri | undefined> => {
+	const { fs } = vscode.workspace;
+
+	// clone Uri
+	let currentUri = vscode.Uri.joinPath(currentUriDir, '.');
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/utils/fs.ts:24]: URIs are immutable so you should not need to clone this

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for taking a look. Added some initial comments. Let me know if you have any questions about them

Can you also please open a new issue to track using node resolution for the tsconfig paths. #131643 was about the referencing the package directory without a `tsconfig.json` at the end, but since this PR adds a lot more functionality we should track that separately

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 2]
```diff
@@ -16,44 +18,61 @@ function mapChildren<R>(node: jsonc.Node | undefined, f: (x: jsonc.Node) => R):
...
 		}
+		const extendsValue = extendsNode.value;
 
-		if (extendsNode.value.startsWith('.')) {
-			return new vscode.DocumentLink(
-				this.getRange(document, extendsNode),
-				vscode.Uri.file(join(dirname(document.uri.fsPath), extendsNode.value + (extendsNode.value.endsWith('.json') ? '' : '.json')))
-			);
-		}
+		const tsconfigUri = await this.getTsconfigPath(Utils.dirname(document.uri), extendsValue);
```
[the remark this answers is not in the document]
Comment (mjbvz) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:45]: With this PR, you are potentially performing a lot of file accesses every time the user types in the file. That's not something we can do

The current implementation also lets you create the file if it does not exist when you click the link

[the remark this answers is not in the document]
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:45]: Okay, in case of relative path we'll always return `.json` path in case if extensionless variant doesn't exist.

But what to do in case if `resolveNodeModulesPath()` returns `undefined`? The previous implementation would just open some path at project's `node_modules`. I don't think user want to create tsconfig at `node_modules`.

I think it makes sense to display error notification to let developer know it. (for example he just forgot to run install command)

**Comment by zardoy** [comment]
> Can you also please open a new issue to track using node resolution for the tsconfig paths. #131643 was about the referencing the package directory without a tsconfig.json at the end, but since this PR adds a lot more functionality we should track that separately

I thought #131643 was about incorrect `extends` resolving algorithm in general, but sure I'll do it in a moment.

File: extensions/typescript-language-features/src/utils/fs.ts [hunk 3]
```diff
@@ -14,3 +15,31 @@ export const exists = async (resource: vscode.Uri): Promise<boolean> => {
...
+		} catch (err) { }
+		if (nodeModulesStat && (nodeModulesStat.type & vscode.FileType.Directory)) {
+			for (const uriCandidate of pathCandidates.map((relativePath) => vscode.Uri.joinPath(nodeModulesUri, relativePath))) {
+				if (await exists(uriCandidate)) {
+					return uriCandidate;
+				}
+			}
+		}
+		// reached the root
+		if (posix.relative(currentUri.path, '/') === '') { return; }
```
[the remark this answers is not in the document]
Comment (zardoy) [reply extensions/typescript-language-features/src/utils/fs.ts]: I can't agree with you. I often use [Monorepo Workspace extension](https://marketplace.visualstudio.com/items?itemName=folke.vscode-monorepo-workspace) to open some package as workspace, I want it resolve correctly in these cases.

Also, git features works in subfolders.

**Comment by zardoy** [comment]
- I changed the implementation and added new internal command.

- I didn't add the nls key, I looked into `./rename.ts` and can see it wasn't defined too.

- I tried to test the command, it opens the tsconfig, but doesn't display error message. Breakpoints works from time to time, so I can't figure out things (everything worked perfectly with previous commit).

    - When i try to open nonexisting file it seems that `window.showTextDocument` and `workspace.openTextDocument` complain about in console (`cannot open ... Unable to resolve nonexistent file`)
    - When it returns `undefined` nothing happens

**Comment by zardoy** [comment]
> Can you also please open a new issue to track using node resolution for the tsconfig paths

I opened #141112, that has a link reference, feel free to request details if I missed something.

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 4]
```diff
@@ -122,5 +175,15 @@ export function register() {
 		languages.map(language =>
 			patterns.map((pattern): vscode.DocumentFilter => ({ language, pattern }))));
 
-	return vscode.languages.registerDocumentLinkProvider(selector, new TsconfigLinkProvider());
+	return vscode.Disposable.from(
+		vscode.commands.registerCommand(OPEN_EXTENDS_LINK_COMMAND_ID, async ({ resourceUri, extendsValue, }: OPEN_EXTENDS_LINK_COMMAND_ARGS) => {
+			const tsconfigPath = await getTsconfigPath(Utils.dirname(resourceUri), extendsValue);
+			if (tsconfigPath === undefined) {
+				void vscode.window.showErrorMessage(localize('openTsconfigExtendsModuleFail', "Failed to resolve {0} as module", extendsValue));
```
Comment (zardoy) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:182]: Added void, as we don't need to wait until user resolve an error.
By adding void we make this line [no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-floating-promises.md) rule aware. I hope it will enabled in https://github.com/microsoft/vscode/issues/128738. 
Or I should throw it instead?

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts]: Also I remember when I was testing it, the link was updating the a HUGE latency. I believe that provider is already called with some latency.

Look, this behavior feels natural: IDE gives a link only when it actually finds a reference, something that points to an actual file.

IMO Creating files via non-existing link is incorrect (I understand that somebody most probably got used to it). There is should be a codefix for appropriate diagnostic instead

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 5]
```diff
@@ -122,5 +175,15 @@ export function register() {
...
 
-	return vscode.languages.registerDocumentLinkProvider(selector, new TsconfigLinkProvider());
+	return vscode.Disposable.from(
+		vscode.commands.registerCommand(openExtendsLinkCommandId, async ({ resourceUri, extendsValue, }: OpenExtendsLinkCommandArgs) => {
+			const tsconfigPath = await getTsconfigPath(Utils.dirname(resourceUri), extendsValue);
+			if (tsconfigPath === undefined) {
+				vscode.window.showErrorMessage(localize('openTsconfigExtendsModuleFail', "Failed to resolve {0} as module", extendsValue));
+				return;
+			}
+			await vscode.workspace.openTextDocument(tsconfigPath);
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:185]: `openTextDocument` opens the text document but you then need to call `showTextDocument` to show it

File: extensions/typescript-language-features/yarn.lock [hunk 6]
```diff
@@ -31,3 +31,8 @@ vscode-nls@^5.0.0:
   version "5.0.0"
   resolved "https://registry.yarnpkg.com/vscode-nls/-/vscode-nls-5.0.0.tgz#99f0da0bd9ea7cda44e565a74c54b1f2bc257840"
   integrity sha512-u0Lw+IYlgbEJFF6/qAqG2d1jQmJl0eyAGJHoAJqr2HT4M2BNuQYSEiSE75f52pXHSJm8AlTjnLLbBFPrdz2hpA==
+
+vscode-uri@^3.0.3:
```
Comment (mjbvz) [inline extensions/typescript-language-features/yarn.lock:35]: Please revert the yarn.lock changes as they cause our CI checks to fail

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 7]
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
...
+			} catch (err) { }
+			if (nodeModulesStat && (nodeModulesStat.type & vscode.FileType.Directory)) {
+				for (const uriCandidate of pathCandidates.map((relativePath) => vscode.Uri.joinPath(nodeModulesUri, relativePath))) {
+					if (await exists(uriCandidate)) {
+						return uriCandidate;
+					}
+				}
+			}
+			// reached the root
+			if (posix.relative(currentUri.path, '/') === '') { return; }
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:158]: This should stop when we reach the root of the workspace, not the root of the user's file system

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 8]
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
...
+ * @returns Returns undefined in case of lack of result while trying to resolve from node_modules
+ */
+const getTsconfigPath = async (baseDirUri: vscode.Uri, extendsValue: string): Promise<vscode.Uri | undefined> => {
+	// Don't take into account a case, where tsconfig might be resolved from the root (see the reference)
+	// e.g. C:/projects/shared-tsconfig/tsconfig.json (note that C: prefix is optional)
+
+	const isRelativePath = ['./', '../'].some(str => extendsValue.startsWith(str));
+	if (isRelativePath) {
+		const absolutePath = vscode.Uri.joinPath(baseDirUri, extendsValue);
+		if (await exists(absolutePath)) { return absolutePath; }
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:128]: [nit] Put body of statements on own line

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 9]
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
...
+		// Will suggest to create a .json variant if it doesn't exist yet
+		return absolutePath.with({
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:130]: For a path `./tsconfig.base.json` that doesn't exist on disk, I believe this will end up trying to create a file called `tsconfig.base.json.json`. Instead you should only add `.json` when the path does not already have a file extension

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 10]
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
...
+			path: `${absolutePath.path}.json`
+		});
+	}
+
+	// Otherwise resolve like a module
+	return resolveNodeModulesPath([
+		extendsValue,
+		`${extendsValue}.json`,
+		`${extendsValue}/tsconfig.json`,
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:139]: We should be able to skip some of these checks entirely. For example if the file already ends in `.json`, there is no purpose to checking `${extendsValue}.json` or `${extendsValue}/tsconfig.json`

Failing these early means we can skip going to disk to check if the file exists

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts [hunk 11]
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
...
+	]);
+
+	async function resolveNodeModulesPath(pathCandidates: string[]): Promise<vscode.Uri | undefined> {
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:142]: [nit] Not a fan of nesting functions definitions like this. Instead move this to the top level of the file and pass in any needed parameters

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 5 above)
Comment (mjbvz) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:185]: Also, you may need to use the `vscode.open` command instead of `openTextDocument` + `showTextDocument`. I believe the command handles the case where the file does not exist and lets users create it if they want (but please test to confirm this)

**Review by mjbvz — CHANGES_REQUESTED** [review]
A few more changes but getting closer!

File: extensions/typescript-language-features/yarn.lock (hunk 6 above)
Comment (zardoy) [reply extensions/typescript-language-features/yarn.lock]: Interesting, whats the point of that workflow?? I reverted the change, but now other actions are failing because of `--frozen-lockfile`

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 11 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:144]: I firstly resolved it because I thought you was talking about moving it above the `return ...` statement. But whats the point of having it defined on top level, since it's being used only in one place? I'm pretty sure this approach reduces readability of your codebase.

TLDR By keeping the definition scoped it is much easier to understand where it's being used. WDYT?

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 7 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts]: Is there any reason for doing this? I don't use `getWorkspaceFolder` and `posix.relative(currentUri.path, currentWorkspacePath.uri.path) === ''` on purpose as this should conform Node resolving algorithm, it was a purpose of this PR!

For example what if someone wants to open `packages/some-package` with structure from https://github.com/microsoft/vscode/pull/141062#issuecomment-1017432148? tssever won't stop at root of the workspace, why we should??

Also, you have already started discussion about that that above.

Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts]: Also, WebStorm works this way..

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 5 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:189]: > `openTextDocument` opens the text document but you then need to call `showTextDocument` to show it

I don't really understand the difference between these, though, even my friend [tried](https://github.com/zardoy/vscode-framework/issues/9#issuecomment-951319097) to explaind it. Nothing worked here except the `vscode.open` command.

However the latter just shows error notification instead of modal, but with desired `create file` button!

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 9 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:131]: You was right! Fixed. I should have better looked at the reference.

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 10 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:140]: But in case if we have `module.json/tsconfig.json` in `node_modules` TS can resolving it by referencing `module.json` from extends.
I'm gonna think about it. I'm probably will be there in a day.

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 11 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:143]: Didn't see a reply so I just moved it like you said. I don't mind of almost any codestyle, sorry for arguing. Btw what nit does mean?

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 7 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts]: I fixed a bug with inifite loop, so now error message displayed correctly. The reason was that `path.relative('C:/', '/')` never gives an empty string.

File: extensions/typescript-language-features/src/languageFeatures/tsconfig.ts (hunk 10 above)
Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:178]: I made impl a little more complicated, now we skip checks for just `node_modules` and check instead full package path. If the package is not present (with structure from https://github.com/microsoft/vscode/pull/141062#issuecomment-1017432148), it won't check all path candidates within first encountered node_modules and instead 5 checks overall will be performed.
Do you have any other ideas on how it can be optimized or on how to make it cleaner?

Comment (zardoy) [reply extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:178]: I wish we had a way to reuse Node's resolving algorithm from somewhere else.

**Comment by zardoy** [comment]
Please review all opened comments when you have time for it. I'll be waiting for reply!

**Comment by mjbvz** [comment]
Thanks! Will be in the next insiders build for testing
